Reworking LMSMetadata package and adding a new lif reader

XLEFReaderForBioformats opened the pull request:
Main changes:
- reworking metadata translation (and class architecture) in the LMSMetadata package, improving metadata translation especially for SP8 and STELLARIS images, also reworking metadata translation for other images (SP5, widefield)
- Adding a new LMSLIFReader for reading .lif files that also uses the LMSMetadata package

In general, metadata translation will be an ongoing topic here, and there might also be images from systems that are not yet fully regarded and where metadata translation needs to be improved. I'll try to fix all major bugs that might be noticed in the context of this PR, but concerning minor bugs and "missing metadata" , I guess I have to discuss with my colleagues how much of that I will be able to fix within this PR.

@dgault as we have discussed before, since you suggested that we could use a DelegateReader here to provide both the existing LIFReader and the new LMSLIFReader to users, I think that's a reasonable approach, so this PR at least won't create a deterioration of existing .lif files reading.
I removed my temporal changes to readers.txt and LIFReader.isThisType(...) for testing so now the LIFReader is the standard reader for .lif files again. Could you let me know how using a delegate reader would look like so both lif readers can be used?

Please let me know if you need any additional information or explanations from my side.

XLEFReaderForBioformats committed:
LOFReader: #3818 LOF file error; [Fatal Error] :1:1: Content is not allowed in prolog.

adapted error messaging

XLEFReaderForBioformats committed:
LIFReader: #3655 Incorrect channel colours, adding LIFReader2 which uses LMS metadata classes (pending discussion).

temporary changes to LIFReader.isThisType() and readers.txt to use LIFReader2 instead of LIFReader for testing

XLEFReaderForBioformats committed:
LMS metadata package: started rework of metadata translation

XLEFReaderForBioformats committed:
Merge remote-tracking branch 'upstream/develop' into xlefreader

XLEFReaderForBioformats committed:
LMS metadata package: reworking stand details, objective, lasers, detectors and filters translation

XLEFReaderForBioformats committed:
LMS metadata package: mapping of detector information from LDM block sequential list to channel detector settings

XLEFReaderForBioformats committed:
LMS metadata package: mapping multiband / filter info from sequential blocks to detector settings and channels

XLEFReaderForBioformats committed:
LMS metadata package: formatting, author/copyright

XLEFReaderForBioformats committed:
rename LIFReader2, LMS metadata package: setting up new structure / architecture (in progress)

XLEFReaderForBioformats committed:
LMS metadata package: reworking channel and dimension metadata handling

XLEFReaderForBioformats committed:
LMSMetadataPackage: setting up usage of new classes

XLEFReaderForBioformats committed:
LMS metadata package: moving channel, dimension and position related info translation

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
LMSMetadata package: i.a. fixing filter and laser settings and physical Z size translation

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
LMSMetadata package: detector read out rate,  laser channel mapping distinguishes between SPX and STELLARIS, fixing channelPrios for LIF/LOF

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
LMS metadata package: removing LMSMetadataExtractor, MetadataStoreInitializer, MetadataTempBuffer

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
LMSMetadata package: ignoring laser settings with intensitydev=0

XLEFReaderForBioformats committed:
LMSMetadata package: bugfix swapXY

XLEFReaderForBioformats committed:
LMSMetadata package: fixing physical z sizes / positions

XLEFReaderForBioformats committed:
LMSMetadata package: renaming Translator

XLEFReaderForBioformats committed:
LMSMetadata package: bugfix Z positions

XLEFReaderForBioformats committed:
LMSMetadata package: bugfixes for LIF XML image nodes finding, empty instruments for images without hardware settings,  empty timestamps,  correct interleaving depending on image format

XLEFReaderForBioformats committed:
LMSMetadata package: refactoring channel priorities / lut color index

XLEFReaderForBioformats committed:
LMSMetadata package: bugfix z position extraction and z position per plane calculation, adding z drive modes

XLEFReaderForBioformats committed:
LMSMetadataPackage: also allow basic metadata translation of SP5 images

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
LMSMetadataPackage: translate filters and exposure times for sequential and non-sequential widefield images

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
LMSMetadataPackage: adding (doc) comments

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
Reverting temporary changes to LIFReader.isThisType(...) for testing

XLEFReaderForBioformats committed:
reverting mapping from .lif files to LMSLIFReader instead of LIFReader

dgault commented:
Thanks @XLEFReaderForBioformats for the effort to get this completed. 

The DelegateReader would be a rather simple new Reader looking similar to https://github.com/ome/bioformats/blob/develop/components/formats-gpl/src/loci/formats/in/ND2Reader.java. There is the longer term option of adding some smarter decision making logic into the DelegateReader but for now the basic version should suffice. This way the 2 readers can be switched through a single API call.

Are there any specific files that would be good to test the new LMSLIFReader?

melissalinkert commented:
@XLEFReaderForBioformats : in order to review this pull request for inclusion, we need to have a better understanding of the testing requirements, including test data as requested in https://github.com/ome/bioformats/pull/4000#issuecomment-1534605538. If we don't hear back within the next week, this pull request will be closed (but you are welcome to re-open it once testing information is added).

XLEFReaderForBioformats commented:
Hi @melissalinkert , sorry for the late reply. We are currently collecting test data for the new LIF reader which we are planning to provide to you, before that we also want to run some in-house tests to get a first idea of what's still missing.

I can hopefully continue my work on this PR in this quarter and will keep you posted on the test data.
We would like to suggest that the new LIF reader is used as standard option, but to avoid any regression, we need to prepare this carefully.

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
temporarily deactivating LIFReader for LMSLifReader testing

XLEFReaderForBioformats committed:
fixing bugs: handling images without hardware settings attachment, wrong pixel rendering, rgb reordering.

also some refactoring

XLEFReaderForBioformats committed:
refactoring: 1 folder for core / OME metadata writers

XLEFReaderForBioformats committed:
refactoring: missing package reference adjustments

XLEFReaderForBioformats committed:
fix xml parsing sequential camera settings

XLEFReaderForBioformats committed:
LMSLIFReader, LOFReader: accessing inverse rgb for correct image (for normal images and tilescans)

XLEFReaderForBioformats committed:
supporting MICA confocal, widefield and widefocal images

XLEFReaderForBioformats committed:
further adjustments for support of MICA CS/WF/WFO images

XLEFReaderForBioformats committed:
fixing bug in timestamp extraction

XLEFReaderForBioformats committed:
supporting different hardware setting setups, adding dye information to channels in MICA images

XLEFReaderForBioformats committed:
reworking handling of hardware setting layouts varying with different imaging systems

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
fixing bug in detector settings translation

XLEFReaderForBioformats committed:
adjusting translation of detectors and detector settings,  also writing dye names to channels

XLEFReaderForBioformats committed:
reworking filter (set) creation, using channel > lightpath > emissionfilterref instead of channel > filtersetref

XLEFReaderForBioformats committed:
major rework of instrument / channel settings translation

bundled extraction of sequential acquisition settings, regarding different laser types, slightly simplified architecture

XLEFReaderForBioformats committed:
remove unused imports

XLEFReaderForBioformats committed:
fix bug in widefield images (main camera setting not set)

XLEFReaderForBioformats committed:
regarding lasers which are not coupled to AOTFs, shutters, light source types and laser power states

XLEFReaderForBioformats committed:
major refactoring of SingleImageTranslator to prepare separate calls of extract() and write() for tile scans

XLEFReaderForBioformats committed:
core and store metadata are now correctly written for image tiles

XLEFReaderForBioformats committed:
fixing timestamp / plane deltaT bug: is now correctly translated

melissalinkert commented:
@XLEFReaderForBioformats : thanks for your continued work here. Is there any update on relevant test data? I understand you may still have some work in progress, but since this is already a very large set of changes we would like to start planning a testing strategy as soon as practical.

XLEFReaderForBioformats commented:
Hi @melissalinkert,

I have collected a new set of test data that I can link here today. It is not yet complete, but I will update it as soon as I have more data at hand.

Is there any other information that you need for planning the testing strategy?
I would try to use the DelegateReader solution as suggested by David, but also we would like to set the new LMSLIFReader as the default option, so that we can reach the majority of users with the new implementation. We're aware that this might require some more careful preparations.

XLEFReaderForBioformats commented:
Hi @melissalinkert ,

here is the test data that I can provide so far. I focused on collecting data from different imaging systems that create different XML layouts.

**Widefield:**

- DM6B Synapse Prime95 (DM6B-Z, AF6000LX)
  - [LIF, XLEF/TIF, XLEF/LOF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/DM6B%20Synapse%20Prime95/DM6B%20Synapse%20Prime95.zip)
- DMi8 K8 LED8 (DMi8, AF6000LX)
  - [LIF](https://webshare3.leica-microsystems.com/fiji/testdata_2024/DMi8_K8_LED8/DMi8%20K8%20LED8%20selection.lif)
  - [XLEF/LOF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/DMi8_K8_LED8/2024_01_22_13_14_09--DMi8%20K8%20LED8%20selection--LOF.zip)
  - [XLEF/TIF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/DMi8_K8_LED8/2024_01_22_13_14_09--DMi8%20K8%20LED8%20selection--TIF.zip)
- M205FCA K3C ITK (M205FCA, AF6000)
  -  [LIF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/M205FCA_K3C_ITK/M205FCA%20K3C%20ITK.zip)
  - [XLEF/LOF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/M205FCA_K3C_ITK/2024_01_22_15_00_47--M205FCA%20K3C%20ITK%20selection--LOF.zip)
  - [XLEF/TIF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/M205FCA_K3C_ITK/2024_01_22_15_00_47--M205FCA%20K3C%20ITK%20selection--TIF.zip)

**MICA:**
- MICA widefield
  - [LIF, XLEF/LOF, XLEF/TIF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/MICA_Widefield/Mica%20Widefield.zip)
- MICA confocal
  - [LIF, XLEF/LOF, XLEF/TIF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/MICA_Confocal/Mica%20Confocal.zip)
- MICA widefocal (confocal + TL)
  - [LIF, XLEF/LOF, XLEF/TIF (zipped)](https://webshare3.leica-microsystems.com/fiji/testdata_2024/MICA_Widefocal/Mica%20Widefocal.zip)

I also provided confocal images in the [first dataset](https://github.com/ome/bioformats/pull/3747#issuecomment-1010087177) (SP8, STELLARIS 5, STELLARIS 8), but I am planning to share SP5 and maybe additional SP8 data as well.


edit @dgault @melissalinkert :
here is an additional SP5 dataset:
- DMI6000 TCS SP5
  - [LIF](https://webshare3.leica-microsystems.com/fiji/testdata_2024/SP5/DMI6000%20TCS%20SP5.lif)
  - [XLEF/TIF (zipped)](https://webshare3.blob.core.windows.net/fiji/testdata_2024/SP5/DMI6000%20TCS%20SP5.zip)

XLEFReaderForBioformats committed:
adding collection path to image name (LIF and XLEF)

XLEFReaderForBioformats committed:
fix build

XLEFReaderForBioformats committed:
bugfix: widefield atl setting empty

XLEFReaderForBioformats committed:
adding translation of objective correction

XLEFReaderForBioformats committed:
avoiding crash for SP5 images without LDM block / atl settings

XLEFReaderForBioformats committed:
fixing laser isActive logic

melissalinkert commented:
Thank you for the update, @XLEFReaderForBioformats. For now, I think if you can let us know when you have finished any planned changes, that would be helpful. If you are planning to use a `DelegateReader`, that should be done before we start reviewing and testing. The current state of this pull request turns off the original LIF reader entirely, so having the `DelegateReader` in place (with either reader as default) would make it much easier to compare the two readers' behavior.

For internal testing, all datasets linked from https://github.com/ome/bioformats/pull/4000#issuecomment-1966113007 are now in `inbox/leica-2024-02-27/`.

XLEFReaderForBioformats commented:
Hi @melissalinkert ,

the only thing that's still pending for me (besides the DelegateReader solution) is the support for SP5 images without LDM block.
I'll finish this until next week hopefully, and then I think we can start the reviewing and testing.

XLEFReaderForBioformats committed:
fixing bug: crash on unformatted objective name

XLEFReaderForBioformats committed:
getting data source type for SP5 images without atl settings

XLEFReaderForBioformats committed:
adding lasers for SP5 images without ATL settings

XLEFReaderForBioformats committed:
getting instrument detectors for SP5 images without atl settings

XLEFReaderForBioformats committed:
adding immersion and correction lookup functions from objective name string (which is not as formatted as assumed)

XLEFReaderForBioformats committed:
confocal settings extraction: major rework to support reading from either ATL settings or from scanner / filter setting records

if scanner / filter settings exist, they are parsed, information is stored in ConfocalSettingRecords for easy access. separating ATL settings and setting records logic

XLEFReaderForBioformats committed:
refactoring folders / names

XLEFReaderForBioformats committed:
adding channel settings with lasers for SP5 images, support of MP lasers / NDD detectors

XLEFReaderForBioformats committed:
SP5 images without atl settings: mapping detectors to detector settings

XLEFReaderForBioformats committed:
adapting shutter check

XLEFReaderForBioformats committed:
parsing shutter info from confocal setting records

XLEFReaderForBioformats committed:
getting correct detector / filter names for SP5 images with and without ATL settings

XLEFReaderForBioformats committed:
adding objective from confocal setting records

XLEFReaderForBioformats commented:
Hi @dgault ,

I am now ready to implement your suggested DelegateReader solution. You linked ND2Reader as an example, is this info still up to date? I found DelegateReader examples in JPEGReader and TiffDelegateReader which look comprehensible to me at first sight, in ND2Reader I couldn't find where the logic for using different readers would be. A clarification would be welcome. :)

edit: nvm, I oriented towards the TiffDelegateReader and it's working ^^ (not sure though if there are any methods missing in LIFDelegateReader that I should add)

XLEFReaderForBioformats committed:
creating LIFDelegateReader so usage of LIFReader and LMSLIFReader can be toggled

dgault commented:
The ND2 Reader example is out of date now as we removed the legacy ND2 reader in Bio-Formats 7.0.0. The older version can be seen [here](https://github.com/ome/bioformats/blob/v6.14.0/components/formats-gpl/src/loci/formats/in/ND2Reader.java), but the other Delegate readers are going to be very similar. Looking quickly at the last commit and it looks to be correct. 

I am going to mark this for inclusion with our nightly repository tests to give us an initial idea of the impact on existing filesets. I will update the PR once the tests have run to see the results.

XLEFReaderForBioformats commented:
okay great, thanks for letting me know! 👍  excited for the results 😄

XLEFReaderForBioformats committed:
correct plane deltaTs and xy positions for RGB images

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
class javadoc + copyright comments

XLEFReaderForBioformats committed:
rename folder, name might be confusing

XLEFReaderForBioformats committed:
correct references

dgault commented:
This PR was included in the CI last night, unfortunately a number of files threw exceptions at the setID stage so the full test suite didn't actually get to run. I have only had a quick look at the failures but a lot of them look to be NullPointerExceptions. Unfortunately the files that are failing are not ones that are public.

The full console output from the test run is at:
https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-test-folder/lastFailedBuild/console

XLEFReaderForBioformats commented:
hi @dgault, okay thanks for sharing, I'll try to fix all the issues where exceptions are thrown.

XLEFReaderForBioformats committed:
increasing robustness against missing / unexpected data, preventing exceptions

XLEFReaderForBioformats commented:
hi @dgault , I tried to address all the exceptions that happened in the tests, trying to increase the robustness against unexpected input.
However, I do not yet understand the root causes of these errors, which are images with data that the reader does not expect.
I understand that you cannot share these image data publicly, however is it maybe possible to make this image data available to me directly?

NicoKiaru commented:
@XLEFReaderForBioformats I'm curious to know whether your request will be approved because I've had exactly the same issue in here:
https://github.com/ome/bioformats/pull/4092

This point has also been raised and answered in the image.sc forum:
https://forum.image.sc/t/pb-when-i-open-czi-stitching-image/87585/24

XLEFReaderForBioformats commented:
@NicoKiaru thanks for sharing, good to know...

@dgault I have a suggestion... I assume that the problems of the new reader with your test data arise from unexpected XML layout and not so much from the image binary data itself. Would it be an option for you to share only the image XML with me privately?

If you have only LIF files available, you could copy the XML using a hex editor, or alternatively you could e.g. use a [LASX Offline](https://www.leica-microsystems.com/products/microscope-software/p/leica-las-x-ls/downloads/) to open LIF files and save them as e.g. XLEF/TIF files, and then it would suffice for me if you sent me the XLIF files, which can be found in metadata folders.
You can also obfuscate any image names, file paths and user-comments in the XML.

Having the image XML, I can create some "fake" image datasets with it for local testing.

Would this be an option?

XLEFReaderForBioformats committed:
using posZ attribute in tile scan info, if available

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
bugfix: in LIF files containing "invalid" memory blocks (e.g. "frame properties"), wrong image offsets are used

mapping of memory blocks to images found in XML via memory block ids, removing offsets of "invalid" memory blocks.
frame properties and other children of image nodes are not considered valid images

XLEFReaderForBioformats committed:
fixing mapping of WLLs to laser line settings

XLEFReaderForBioformats commented:
hi @dgault @melissalinkert,
could you possibly support me with the latest build errors? my changes built locally, and for some reason I cannot see the build logs of the failed builds (it only displays "this job failed")...

dgault commented:
@XLEFReaderForBioformats, Im not sure why those jobs failed in that way but I re-triggered them again and they should now all be green

XLEFReaderForBioformats committed:
translating Z range / position for SP5 images without ATL settings

XLEFReaderForBioformats committed:
bug fix zEnd calculation

XLEFReaderForBioformats commented:
@dgault I'd like to ask if there are any updates regarding the test data? It would be helpful for me to know at least if you are planning to share any data with me or not, right now it is a bit unclear to me how to proceed with this PR.

dgault commented:
Hi @XLEFReaderForBioformats, the next step will be to reinclude the PR in the nightly CI, testing the new reader rather than the legacy reader. If we can get a full run of the test suite then we will at least be able to identify the number of files which are causing failures and judge the scale of the impact. From there we can decide how best to proceed. 

I will re-include the PR now so we can have the new reader in the nightly CI for the purposes of the testing.

dgault commented:
Hi @XLEFReaderForBioformats, as a follow up, the PR has been included in our test suite for a number of days and we are seeing failures for a number of files, though most are for the same exception. You can see the LIF failures here:
https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-test-folder/7512/consoleFull

```
   [testng] java.lang.IndexOutOfBoundsException: Index 751 out of bounds for length 13
   [testng] 	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
   [testng] 	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
   [testng] 	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
   [testng] 	at java.base/java.util.Objects.checkIndex(Objects.java:385)
   [testng] 	at java.base/java.util.ArrayList.get(ArrayList.java:427)
   [testng] 	at loci.formats.in.LMSLIFReader.initFile(LMSLIFReader.java:461)
   [testng] 	at loci.formats.FormatReader.setId(FormatReader.java:1480)
   [testng] 	at loci.formats.in.LIFDelegateReader.setId(LIFDelegateReader.java:37)
   [testng] 	at loci.formats.ImageReader.setId(ImageReader.java:864)
```

The failures seem to be caused by the following line when retrieving the offsets:
https://github.com/ome/bioformats/pull/4000/files#diff-79e103bcccaa0d842ec61c1bf11e438d3ace6607e364135a7e08b3947f2dec7cR461

The following sample files which failed are available for testing from [here](https://downloads.openmicroscopy.org/images/Leica-LIF/). These should be good samples to test for the IndexOutOfBoundsException
- michael/PR2729_frameOrderCombinedScanTypes.lif
- imagesc-30856/20191025 Test FRET 585. 423, 426.lif
- seanwarren/150519_FRAP_test_ROIs_chromagreen/150519_FRAP_test_ROIs_chromagreen.lif

XLEFReaderForBioformats committed:
Merge branch 'ome:develop' into xlefreader

XLEFReaderForBioformats committed:
fixing IndexOutOfBoundsException in LMSLIFReader.initFile()

XLEFReaderForBioformats committed:
fixing NullPointerException in ConfocalSettingsFromSettingRecordsExtractor.createChannelSettings(..)

XLEFReaderForBioformats commented:
Hi @dgault,

sorry for the long delay on my side! After a longer phase of release support that I had to prioritize, I can now hopefully return to finishing this PR, and I'm really looking forward to it!

I fixed two bugs that led to the exceptions in the test log that you kindly provided.

Since it's been a few months since the last update from my side, I am unsure as to how we best proceed now. Would you simply include the PR in your test suite again?

XLEFReaderForBioformats commented:
Hi @melissalinkert ,

I'd like to kindly ask again how we can proceed with this PR, as I fixed all issues that were found in your automated tests.
Is there anything else missing before this PR can be merged?

melissalinkert commented:
@XLEFReaderForBioformats : we're currently evaluating what is needed to proceed with reviewing this pull request. Given the size and scope of changes, we expect review and testing to require a substantial amount of time and effort. Due to holidays and other priorities, it will likely be after the new year before we have an update here.

Please note too that we have had some staffing changes in the last few months, details of which are mentioned [here](https://forum.image.sc/t/bio-formats-update/101734).

XLEFReaderForBioformats commented:
Hi @melissalinkert , are there any updates on your end?

melissalinkert commented:
My apologies for the delay in responding.

Thanks again to everyone for all the work & discussion that has gone into this PR. We truly appreciate the effort it takes to develop a new reader and the shared goal of improving Bio-Formats.

After careful internal discussions, we’ve decided that we still aren’t able to take responsibility for this contribution. The proposed changes are large, and replacing the existing reader would mean losing a lot of historical knowledge embedded in the original implementation. Given the scope and impact, we aren’t in a position to take on the adoption and long-term maintenance of this reader as part of the core Bio-Formats library.

That said, this does open up an exciting opportunity to improve how to enable the testing of external readers! We’ll continue to support externally hosted readers as an extension point via readers.txt.

Recent PRs and related conversations on the public image.sc forum like https://forum.image.sc/t/would-anyone-have-a-palm-czi-example-file/85900 and https://forum.image.sc/t/pb-when-i-open-czi-stitching-image/87585 highlighted the importance of public permissively licensed datasets to support Bio-Formats development and allow contributions. Historical data received without a public license cannot be shared, but as per [OME contributing policy](https://ome-contributing.readthedocs.io/en/latest/third-party-policy.html#receipt-of-sample-data), all new data is strongly encouraged to be submitted publicly using the [Zenodo Bio-Formats community](https://zenodo.org/communities/bio-formats). Owners of previously submitted data, including vendors of the data, who would like to make their data public should resubmit a version using the Zenodo community and we will update our curated OME QA repository and [public download page](https://downloads.openmicroscopy.org/images/) to reflect the new status.

We really appreciate your contributions and look forward to collaborating on ways to enhance external reader testing.

XLEFReaderForBioformats commented:
Hi @melissalinkert ,

thank you for your response!

Would it be more manageable for your team to test this PR if I removed the changes made to the LIFReader?
Then the PR would only include the improvements that I added for the LMSMetadata package, which affects the XLEFReader and LOFReader.

Best regards
Constanze
